[BUG] Trace is `not opening` on specific broken locator

<!-- ⚠️⚠️ Do not delete this template ⚠️⚠️ -->

<!-- 🔎 Search existing issues to avoid creating duplicates. -->
<!-- 🧪 Test using the latest Playwright release to see if your issue has already been fixed -->
<!-- 💡 Provide enough information for us to be able to reproduce your issue locally -->

### System info
- Playwright Version: [v1.31.2]
- Operating System: [All]
- Browser: [Chromium ( maybe all )]
- Other info:

### Source code

- [x] I provided exact source code that allows reproducing the issue locally.

<!-- For simple cases, please provide a self-contained test file along with the config file -->
<!-- For larger cases, you can provide a GitHub repo you created for this issue -->
<!-- If we can not reproduce the problem locally, we won't be able to act on it -->
<!-- You can still file without the exact code and we will try to help, but if we can't repro, it will be closed -->


**Config file**

```js
// playwright.config.ts

  trace: "retain-on-failure",

```

**Test file (self-contained)**

```js
test.describe("Sample test", () => {
    test('How to kill trace', async ({page}) => {
        await page.goto('https://playwright.dev/');

        await page.locator("[class*=github-btn] a]").click()
    })
})
```

**Steps**
- [Run the test]
- test fails
- open trace

**Expected**

trace opened, with errors from the console

**Actual**

blank screen in place of a trace, only after opening browser's console one could see a broken locator

```js
index.f6c909e6.js:40 Error: Unexpected token "]" while parsing selector "[class*=github-btn] a]"
    at h (index.f6c909e6.js:41:3984)
    at A (index.f6c909e6.js:41:5695)
    at M (index.f6c909e6.js:41:4966)
    at S (index.f6c909e6.js:41:4664)
    at T (index.f6c909e6.js:41:4572)
    at HS (index.f6c909e6.js:41:5837)
    at index.f6c909e6.js:41:6507
    at Array.map (<anonymous>)
    at W0 (index.f6c909e6.js:41:6383)
    at B0 (index.f6c909e6.js:41:10836)
Cc @ index.f6c909e6.js:40
index.f6c909e6.js:41 Uncaught (in promise) Error: Unexpected token "]" while parsing selector "[class*=github-btn] a]"
    at h (index.f6c909e6.js:41:3984)
    at A (index.f6c909e6.js:41:5695)
    at M (index.f6c909e6.js:41:4966)
    at S (index.f6c909e6.js:41:4664)
    at T (index.f6c909e6.js:41:4572)
    at HS (index.f6c909e6.js:41:5837)
    at index.f6c909e6.js:41:6507
    at Array.map (<anonymous>)
    at W0 (index.f6c909e6.js:41:6383)
    at B0 (index.f6c909e6.js:41:10836)
```

--- a/packages/trace-viewer/src/ui/actionList.tsx
+++ b/packages/trace-viewer/src/ui/actionList.tsx
@@ -60,7 +60,7 @@ const renderAction = (
   revealConsole: () => void
 ) => {
   const { errors, warnings } = modelUtil.stats(action);
-  const locator = action.params.selector ? asLocator(sdkLanguage || 'javascript', action.params.selector) : undefined;
+  const locator = action.params.selector ? asLocator(sdkLanguage || 'javascript', action.params.selector, false /* isFrameLocator */, true /* playSafe */) : undefined;
 
   let time: string = '';
   if (action.endTime)

--- a/packages/trace-viewer/src/ui/callTab.tsx
+++ b/packages/trace-viewer/src/ui/callTab.tsx
@@ -95,7 +95,7 @@ function propertyToString(event: ActionTraceEvent, name: string, value: any, sdk
   if ((name === 'value' && isEval) || (name === 'received' && event.method === 'expect'))
     value = parseSerializedValue(value, new Array(10).fill({ handle: '<handle>' }));
   if (name === 'selector')
-    return { text: asLocator(sdkLanguage || 'javascript', event.params.selector), type: 'locator', name: 'locator' };
+    return { text: asLocator(sdkLanguage || 'javascript', event.params.selector, false /* isFrameLocator */, true /* playSafe */), type: 'locator', name: 'locator' };
   const type = typeof value;
   if (type !== 'object' || value === null)
     return { text: String(value), type, name };

--- a/packages/trace-viewer/src/ui/snapshotTab.tsx
+++ b/packages/trace-viewer/src/ui/snapshotTab.tsx
@@ -265,7 +265,7 @@ export const InspectModeController: React.FunctionComponent<{
       recorder = new Recorder(injectedScript, {
         async setSelector(selector: string) {
           recorder!.setUIState({ mode: 'none', language: sdkLanguage, testIdAttributeName });
-          setHighlightedLocator(asLocator('javascript', selector, false));
+          setHighlightedLocator(asLocator('javascript', selector, false /* isFrameLocator */, true /* playSafe */));
         }
       });
       win._recorder = recorder;

--- a/tests/library/trace-viewer.spec.ts
+++ b/tests/library/trace-viewer.spec.ts
@@ -689,6 +689,19 @@ test('should include requestUrl in route.fulfill', async ({ page, runAndTrace, b
   await expect(callLine.getByText('requestUrl')).toContainText('http://test.com');
 });
 
+test('should not crash with broken locator', async ({ page, runAndTrace, server }) => {
+  test.info().annotations.push({ type: 'issue', description: 'https://github.com/microsoft/playwright/issues/21832' });
+  const traceViewer = await runAndTrace(async () => {
+    try {
+      await page.locator("[class*=github-btn] a]").click();
+    } catch (e) {
+    }
+  });
+  await expect(traceViewer.page).toHaveTitle('Playwright Trace Viewer');
+  const header = traceViewer.page.getByText('Playwright', { exact: true });
+  await expect(header).toBeVisible();
+});
+
 test('should include requestUrl in route.continue', async ({ page, runAndTrace, server }) => {
   await page.route('**/*', route => {
     route.continue({ url: server.EMPTY_PAGE });

--- a/tests/library/trace-viewer.spec.ts
+++ b/tests/library/trace-viewer.spec.ts
@@ -693,7 +693,7 @@ test('should not crash with broken locator', async ({ page, runAndTrace, server 
   test.info().annotations.push({ type: 'issue', description: 'https://github.com/microsoft/playwright/issues/21832' });
   const traceViewer = await runAndTrace(async () => {
     try {
-      await page.locator("[class*=github-btn] a]").click();
+      await page.locator('[class*=github-btn] a]').click();
     } catch (e) {
     }
   });

